linalg/constraints.cpp

```diff
@@ -0,0 +1,893 @@
+                                     Array<int>& secondary_dofs)
+   :
+   ConstrainedSolver(A.GetComm(), A, B),
+   hA(A),
+   prec(nullptr)
+{
+   MFEM_VERIFY(secondary_dofs.Size() == B.Height(),
+               "Wrong number of dofs for elimination!");
+   Array<int> lagrange_dofs(secondary_dofs.Size());
+   for (int i = 0; i < lagrange_dofs.Size(); ++i)
+   {
+      lagrange_dofs[i] = i;
+   }
+   eliminators.Append(new Eliminator(B, lagrange_dofs, primary_dofs,
+                                     secondary_dofs));
+   projector = new EliminationProjection(hA, eliminators);
+}
+
+EliminationSolver::EliminationSolver(HypreParMatrix& A, SparseMatrix& B,
+                                     Array<int>& constraint_rowstarts)
+   :
+   ConstrainedSolver(A.GetComm(), A, B),
+   hA(A),
+   prec(nullptr)
+{
+   if (!B.Empty())
+   {
+      int * I = B.GetI();
+      int * J = B.GetJ();
+      double * data = B.GetData();
+
+      for (int k = 0; k < constraint_rowstarts.Size() - 1; ++k)
+      {
+         int constraint_size = constraint_rowstarts[k + 1] -
+                               constraint_rowstarts[k];
+         Array<int> lagrange_dofs(constraint_size);
+         Array<int> primary_dofs;
+         Array<int> secondary_dofs(constraint_size);
+         secondary_dofs = -1;
+         // loop through rows, identify one secondary dof for each row
+         for (int i = constraint_rowstarts[k]; i < constraint_rowstarts[k + 1]; ++i)
+         {
+            lagrange_dofs[i - constraint_rowstarts[k]] = i;
+            for (int jptr = I[i]; jptr < I[i + 1]; ++jptr)
+            {
+               int j = J[jptr];
+               double val = data[jptr];
+               if (std::abs(val) > 1.e-12 && secondary_dofs.Find(j) == -1)
+               {
+                  secondary_dofs[i - constraint_rowstarts[k]] = j;
+                  break;
+               }
+            }
+         }
+         // loop through rows again, assigning non-secondary dofs as primary
+         for (int i = constraint_rowstarts[k]; i < constraint_rowstarts[k + 1]; ++i)
+         {
+            MFEM_ASSERT(secondary_dofs[i - constraint_rowstarts[k]] >= 0,
+                        "Secondary dofs don't match rows!");
+            for (int jptr = I[i]; jptr < I[i + 1]; ++jptr)
+            {
+               int j = J[jptr];
+               if (secondary_dofs.Find(j) == -1)
+               {
+                  primary_dofs.Append(j);
+               }
+            }
+         }
+         primary_dofs.Sort();
+         primary_dofs.Unique();
+         eliminators.Append(new Eliminator(B, lagrange_dofs, primary_dofs,
+                                           secondary_dofs));
+      }
+   }
+   projector = new EliminationProjection(hA, eliminators);
+}
+
+void EliminationSolver::PrimalMult(const Vector& rhs, Vector& sol) const
+{
+   if (!prec)
+   {
+      prec = BuildPreconditioner();
+   }
+   IterativeSolver * krylov = BuildKrylov();
```

Comment:
It seems to me that it makes sense to only build the krylov solver once and then keep it alive on the class. I noticed you did this a couple of different places, and my suggestion would be change it in those other places as well.

Assuming these set operations aren't dynamic, you could probably also get away with only do them once as well.